src/llm.py

```diff
@@ -0,0 +1,317 @@
+            event_num, event_topic = next_event.name.split(":", maxsplit=1)
+            result.append(f"Next bust event: {event_num.strip()}")
+            result.append(f"Next bust topic: {event_topic.strip()}")
+            result.append(f"Next bust time: {next_event.start_time.strftime('%b %d')}")
+        else:
+            result.append(f"Next event: {next_event.name}")
+            result.append(f"Next event time: {next_event.start_time.strftime('%b %d')}")
+
+        result.append(f"Today's date: {datetime.date.today().strftime('%b %d')}")
+
+    user = get_name(message.author)
+
+    # Role-based info
+    if isinstance(message.author, Member):
+        roles = {role.name for role in message.author.roles}
+
+        # Provide champion info
+        champ = [
+            ("Defending Champion", "first"),
+            ("Runner-up", "second"),
+            ("Bronzer", "third"),
+        ]
+        for role, place in champ:
+            if role in roles:
+                result.append(f"{user}'s place last bust: {place}")
+                break
+
+    result.append(f"Talking to: {user}")
+    return result
+
+
+# Count tokens in a string
+def token_count(data: str) -> int:
+    return len(encoding.encode(data))
+
+
+# Replace Discord-style mentions with names
+def substitute_mentions(message: Message) -> str:
+    content = message.content
+    for user in message.mentions:
+        content = content.replace(user.mention, get_name(user))
+    for channel in message.channel_mentions:
+        content = content.replace(channel.mention, channel.name)
+    for role in message.role_mentions:
+        content = content.replace(role.mention, role.name)
+    return content
+
+
+# Fetch message content from history up to a certain token allowance
+async def fetch_history(
+    token_limit: int, speaking_turn_limit: int, message: Message
+) -> List[Tuple[str, bool]]:
+    total_tokens = 0
+    # A list of (str, bool) tuples containing:
+    #     - "{message author}: {message content}"
+    #     - True if the the message was sent by the bot user, False otherwise
+    history = []
+    idx = 0
+    one_hour = datetime.timedelta(hours=1)
+    seen_message = False
+    speaking_turn_count = 0
+    last_author = None
+    async for msg in message.channel.history():
+        # Skip messages newer than what we're replying to
+        seen_message = seen_message or (msg == message)
+        if not seen_message:
+            continue
+
+        # Skip if message is disallowed
+        if disallowed_message(msg):
+            continue
+
+        # Don't include messages which are both more than 3 back and an hour old
+        # as the discussion has likely moved on.
+        idx += 1
+        timezone = message.created_at.tzinfo
+        now = datetime.datetime.now(timezone)
+        if idx > 3 and (now - message.created_at) > one_hour:
+            break
+
+        # Break if maximum number of speaking turns has been reached
```

Comment:
```suggestion
        # Break if maximum number of speaking turns has been reached
        # We use turns instead of a simple message count as some people like to
        # break up their thoughts into many separate messages.
```